src/sync.rs

```diff
@@ -554,6 +637,133 @@ where
+/// blocking manner because many C API calls can trigger the execution of
+/// arbitrary Python code.
+///
+/// Only available on Python 3.14 and newer.
+#[cfg(all(Py_3_14, not(Py_LIMITED_API)))]
+#[cfg_attr(not(Py_GIL_DISABLED), allow(unused_variables))]
+pub fn with_critical_section_mutex<'py, 'a, F, R, T>(
+    _py: Python<'py>,
+    mutex: &'a PyMutex<T>,
+    f: F,
+) -> R
+where
+    F: FnOnce(EnteredCriticalSection<'a, T>) -> R,
+{
+    #[cfg(Py_GIL_DISABLED)]
+    {
+        let mut guard = CSGuard(unsafe { std::mem::zeroed() });
+        unsafe { crate::ffi::PyCriticalSection_BeginMutex(&mut guard.0, &mut *mutex.mutex.get()) };
+        f(EnteredCriticalSection(&mutex.data))
+    }
+    #[cfg(not(Py_GIL_DISABLED))]
+    {
+        f(EnteredCriticalSection(&mutex.data))
+    }
+}
+
+/// Executes a closure with a Python critical section held on two `PyMutex` instances.
+///
+/// Simultaneously acquires the mutexes `m1` and `m2` and holds them
+/// until the closure `f` is finished. The mutexes may be
+/// temporarily released and re-acquired if the closure calls back into the
+/// interpreter in a manner that would block.  This is similar to how the GIL
+/// can be released during blocking calls. See the safety notes below for
+/// caveats about releasing critical sections.
+///
+/// This is structurally equivalent to the use of the paired
+/// Py_BEGIN_CRITICAL_SECTION2_MUTEX and Py_END_CRITICAL_SECTION2 C-API macros.
+///
+/// A no-op on GIL-enabled builds, where the critical section API is exposed as
+/// a no-op by the Python C API.
+///
+/// # Safety
+///
+/// Provides weaker locking guarantees than traditional locks, but can in some
+/// cases be used to provide guarantees similar to the GIL without the risk of
+/// deadlocks associated with traditional locks.
+///
+/// The caller must ensure the closure cannot implicitly release the critical
+/// section. If a multithreaded program calls back into the Python interpreter
+/// in a manner that would cause the critical section to be released, the
+/// `PyMutex` will be released and the resource protected by the `PyMutex` may
+/// be read from or modified by another thread. Concurrent modifications are
+/// impossible, but races are possible and the state of an object may change
+/// "underneath" a suspended thread in possibly surprising ways. Note that many
+/// operations on Python objects may call back into the interpreter in a
+/// blocking manner because many C API calls can trigger the execution of
+/// arbitrary Python code.
+///
+/// Only available on Python 3.14 and newer.
+#[cfg(all(Py_3_14, not(Py_LIMITED_API)))]
+#[cfg_attr(not(Py_GIL_DISABLED), allow(unused_variables))]
+pub fn with_critical_section_mutex2<'py, 'a, F, R, T1, T2>(
+    _py: Python<'py>,
+    m1: &'a PyMutex<T1>,
+    m2: &'a PyMutex<T2>,
+    f: F,
+) -> R
+where
+    F: FnOnce(EnteredCriticalSection<'a, T1>, EnteredCriticalSection<'a, T2>) -> R,
```

Comment:
Similar to the above


```suggestion
    F: for<'section> FnOnce(EnteredCriticalSection<'section, T1>, EnteredCriticalSection<'section, T2>) -> R,
```